### DISCUSSION TITLE
Eliminate global OptSearchReverse

### DISCUSSION THREAD

**Discussion by flatcap** [neomutt/neomutt#3997] [Development]
This looks like it should be trivial.

`OptSearchReverse` is used in two functions, `mutt_search_command()` and `mutt_search_alias_command()`.
They are independent, so it should be possible to replace `OptSearchReverse` with a local variable.

Once done, check that searching works correctly in the Index and Address Book (Aliases).
i.e.
- `<search>` searches forwards
- `<search-reverse>` searches backwards
- `<search-next>` continues searching in the same direction
- `<search-opposite>` searches in the opposite direction

Here are the places where `OptSearchReverse` is referenced.

```
pattern/pattern.c:465:        OptSearchReverse = false;
pattern/pattern.c:467:        OptSearchReverse = true;
pattern/pattern.c:516:        int incr = OptSearchReverse ? -1 : 1;
pattern/pattern.c:631:        OptSearchReverse = false;
pattern/pattern.c:633:        OptSearchReverse = true;
pattern/pattern.c:680:        int incr = OptSearchReverse ? -1 : 1;
```

**Comment by roccoblues** [comment]
With "local variable" you mean a specific global per function in `pattern.c` right? Because the functions still rely on the state being kept across function calls.

[Here](https://github.com/neomutt/neomutt/blob/main/pattern/pattern.c#L464) we set it, but [only when we "start the search"](https://github.com/neomutt/neomutt/blob/main/pattern/pattern.c#L453). Subsequent `<search-next>` or `<search-opposite>` [then use it](https://github.com/neomutt/neomutt/blob/main/pattern/pattern.c#L516).

**Comment by flatcap** [comment]
Ah, yes.  Now I see.

Could we instead, pass in the value from the three `op_search()` functions? (in index, postponed, alias)

**Reply by roccoblues** [reply]
> Could we instead, pass in the value from the three op_search() functions? (in index, postponed, alias)

You mean keep the global but pass something like `reverseSearch` it into `mutt_search_command()` and `mutt_search_alias_command()` so that at least they don't use it?

I can look into that. But I'm not sure if it makes sense without also removing their reliance on `OP_SEARCH*`.

**Reply by flatcap** [reply]
Currently, I think that searching in one dialog affects the others.
I was thinking (again, without looking deeply into the code :-),
that we could have a `bool search_direction` in the `IndexPrivateData` (and similar for postpone, alias).

**Comment by roccoblues** [comment]
Added a first idea here: https://github.com/neomutt/neomutt/pull/4017

**Comment by flatcap** [comment]
Completed in: 0b51b8351 remove search globals
